### ISSUE TITLE
ppw3, specific pdf takes significantly longer to load

### ISSUE THREAD

**Issue by luutuankiet** [koreader/koreader#10875]
* KOReader version: v2023.06.1
* Device: kindle PPW3

#### Issue
Koreader takes longer to load (~5min) this pdf which I also experienced with a Boox device in the below issue; though I suspect these are not 100% the same bug since in the Boox issue the device hung much longer sometimes froze completely.
https://github.com/koreader/koreader/issues/10130

#### Steps to reproduce
Open this pdf and observe
https://github.com/ms2ag16/Books/blob/master/Kimball_The-Data-Warehouse-Toolkit-3rd-Edition.pdf

##### `crash.log` 
inspecting the `crash.log` at 09/04/23-14:46:43 didn't show any particular loops - koreader simply logged 2 minutes of loading the document. Loading such pdf with the PPW3's native kindle app went fine which took a couple of seconds.

If in the end we conclude that it's just the document is at fault then I hope we can identify some of its attributes to avoid getting similar documents altogether.

<details>
  <summary>crash.log</summary>

```
---------------------------------------------
                launching...
  _  _____  ____                _
 | |/ / _ \|  _ \ ___  __ _  __| | ___ _ __
 | ' / | | | |_) / _ \/ _` |/ _` |/ _ \ '__|
 | . \ |_| |  _ <  __/ (_| | (_| |  __/ |
 |_|\_\___/|_| \_\___|\__,_|\__,_|\___|_|

 It's a scroll... It's a codex... It's KOReader!

 [*] Current time: 09/01/23-14:03:41
 [*] Version: v2023.06.1

ffi.load: libs/libutf8proc.so.2
ffi.load: blitbuffer
09/01/23-14:03:42 INFO  initializing for device KindlePaperWhite3 
09/01/23-14:03:42 INFO  framebuffer resolution: {
  h = 1448,
  w = 1072
} --[[table: 0x403ae128]] 
[ko-input] Forked off fake event generator (pid: 13489)
ffi.load: libs/libmupdf.so
ffi.load: libs/libwrap-mupdf.so
ffi.load: libs/libfreetype.so.6
ffi.load: libs/libharfbuzz.so.0
ffi.load: libs/libzstd.so.1
09/01/23-14:03:43 INFO  opening file /mnt/us/documents/main/pro/programming/Kimball_The-Data-Warehouse-Toolkit-3rd-Edition.pdf 
ffi.load: libs/liblept.so.5
ffi.load: libs/libk2pdfopt.so.2
09/01/23-14:05:40 INFO  Inhibiting user input 
09/01/23-14:05:40 INFO  Loading plugins from directory: plugins 
ffi.load: sqlite3
09/01/23-14:05:40 INFO  setting zoom mode to contentwidth 
09/01/23-14:05:42 INFO  Restoring user input handling 
09/01/23-14:05:44 INFO  Inhibiting user input 
09/01/23-14:05:46 INFO  Restoring user input handling 
09/01/23-14:05:52 INFO  Inhibiting user input 
09/01/23-14:16:25 INFO  Restoring user input handling 
09/01/23-14:18:39 INFO  Terminal: spawn done 
09/01/23-14:20:09 INFO  Terminal: spawn done 
09/01/23-14:47:00 INFO  Inhibiting user input 
09/01/23-15:57:20 WARN  Kindle unscheduled wakeup 
09/01/23-16:10:30 WARN  Kindle unscheduled wakeup 
09/02/23-13:11:06 WARN  Kindle unscheduled wakeup 
09/02/23-13:45:36 INFO  Restoring user input handling 
09/02/23-13:46:00 INFO  writing document to /mnt/us/documents/main/pro/programming/Kimball_The-Data-Warehouse-Toolkit-3rd-Edition.pdf 
09/02/23-13:46:00 INFO  opening file /mnt/us/documents/main/soul/non-fiction/Sapiens - Yuval noah Harari.epub 
09/02/23-13:46:01 ERROR failed to register crengine font: cannot register font <./fonts/bookerly-cufonfonts/Bookerly LCD Italic.ttf> 
09/02/23-13:46:01 ERROR failed to register crengine font: cannot register font <./fonts/bookerly-cufonfonts/Bookerly Light Italic.ttf> 
09/02/23-13:46:01 ERROR failed to register crengine font: cannot register font </usr/java/lib/fonts/Bookerly-Bold.ttf> 
09/02/23-13:46:01 ERROR failed to register crengine font: cannot register font </usr/java/lib/fonts/Bookerly-BoldItalic.ttf> 
09/02/23-13:46:01 ERROR failed to register crengine font: cannot register font </usr/java/lib/fonts/Bookerly-Italic.ttf> 
09/02/23-13:46:01 ERROR failed to register crengine font: cannot register font </usr/java/lib/fonts/Bookerly-Regular.ttf> 
09/02/23-13:46:01 ERROR failed to register crengine font: cannot register font </usr/java/lib/fonts/BookerlyDisplay-Bold.ttf> 
09/02/23-13:46:01 ERROR failed to register crengine font: cannot register font </usr/java/lib/fonts/BookerlyDisplay-BoldItalic.ttf> 
09/02/23-13:46:01 ERROR failed to register crengine font: cannot register font </usr/java/lib/fonts/BookerlyDisplay-Italic.ttf> 
09/02/23-13:46:01 ERROR failed to register crengine font: cannot register font </usr/java/lib/fonts/BookerlyDisplay-Regular.ttf> 
CRE: font NotoSerifKR regular: updated weight from 500 to 400
CRE: font NotoSerifKR bold: updated weight from 900 to 700
CRE: font PMN Caecilia LT regular: updated weight from 500 to 400
CRE: font NotoSansKR bold: updated weight from 900 to 700
09/02/23-13:46:01 INFO  Inhibiting user input 
09/02/23-13:46:03 INFO  Restoring user input handling 
09/02/23-13:46:19 INFO  Inhibiting user input 
09/02/23-13:46:19 INFO  Restoring user input handling 
09/02/23-13:47:48 INFO  Inhibiting user input 
09/02/23-13:47:48 INFO  Restoring user input handling 
09/02/23-13:48:34 INFO  opening file /mnt/us/documents/main/soul/non-fiction/Sapiens - Yuval noah Harari.epub 
09/02/23-13:48:34 INFO  Inhibiting user input 
09/02/23-13:48:35 INFO  Restoring user input handling 
09/02/23-13:53:07 INFO  Inhibiting user input 
09/02/23-14:03:45 INFO  Restoring user input handling 
09/02/23-14:04:04 INFO  Inhibiting user input 
09/02/23-14:04:04 INFO  Restoring user input handling 
09/02/23-14:05:17 INFO  opening file /mnt/us/documents/main/soul/non-fiction/Sapiens - Yuval noah Harari.epub 
09/02/23-14:05:17 INFO  Inhibiting user input 
09/02/23-14:05:18 INFO  Restoring user input handling 
09/02/23-14:15:23 INFO  Terminal: spawn done 
09/02/23-14:15:23 INFO  Inhibiting user input 
09/02/23-14:15:24 INFO  Restoring user input handling 
09/02/23-14:25:23 INFO  Inhibiting user input 
09/02/23-14:25:24 INFO  Restoring user input handling 
09/02/23-14:31:34 INFO  Inhibiting user input 
09/02/23-17:38:25 INFO  Restoring user input handling 
09/02/23-17:38:33 INFO  Inhibiting user input 
09/03/23-12:36:55 INFO  Restoring user input handling 
09/03/23-13:06:57 INFO  Inhibiting user input 
09/03/23-17:32:14 WARN  Kindle unscheduled wakeup 
09/03/23-17:36:22 WARN  Kindle unscheduled wakeup 
09/03/23-17:38:24 WARN  Kindle unscheduled wakeup 
09/03/23-17:50:24 WARN  Kindle unscheduled wakeup 
09/03/23-18:00:45 WARN  Kindle unscheduled wakeup 
09/03/23-18:02:45 WARN  Kindle unscheduled wakeup 
09/03/23-18:05:19 WARN  Kindle unscheduled wakeup 
09/04/23-12:24:47 INFO  Restoring user input handling 
09/04/23-12:55:24 INFO  Inhibiting user input 
09/04/23-12:55:27 INFO  Restoring user input handling 
09/04/23-12:57:23 INFO  Inhibiting user input 
09/04/23-14:46:02 INFO  Restoring user input handling 
09/04/23-14:46:43 INFO  opening file /mnt/us/documents/main/pro/programming/Kimball_The-Data-Warehouse-Toolkit-3rd-Edition.pdf 
09/04/23-14:48:40 INFO  Inhibiting user input 
09/04/23-14:48:40 INFO  setting zoom mode to contentwidth 
09/04/23-14:48:42 INFO  Restoring user input handling 
09/04/23-14:51:00 INFO  no dialog left to show 
09/04/23-14:51:00 INFO  Tearing down UIManager with exit code: 0 
[ko-input] Closed input device with fd: 10 @ idx: 1
[ko-input] Closed input device with fd: 9 @ idx: 0
lipc-wait-event exited normally with status: 0
```
</details>

**Comment by Frenzie** [comment]
Interesting, that PDF loads slowly in MuPDF, also 1.21, and it takes several seconds even on my PC. That's unusual, since normally it's not slower than Poppler and frequently faster, but Poppler does this one instantaneously. But yeah, that makes it a document issue in practical terms. I don't have MuPDF 1.22 or 1.23 at hand right this moment though; maybe it's been resolved since.

My suspicion would be it's related to the quite sizable PNG with transparency on the front page.

**Comment by luutuankiet** [comment]
worked as expected after I followed instruction above to remove the 1st page.

**Comment by Frenzie** [comment]
Another alternative is something like this: `mutool convert -O compress -o Kimball_The-Data-Warehouse-Toolkit-3rd-Edition-mupdf.pdf Kimball_The-Data-Warehouse-Toolkit-3rd-Edition.pdf`. It loads quickly, it's slightly smaller, and I don't think it should display any different. The command itself will take roughly forever for the same reason though.

If I check the original it seems to be regular gzip (same as after conversion above), which should normally be quite speedy. I don't really see anything that stands out to me as odd, but I'm not an expert.

```
50434 0 obj<</First 2272/Length 2865/Filter/FlateDecode/N 200/Type/ObjStm>>stream
50435 0 obj<</First 2150/Length 1021/Filter/FlateDecode/N 200/Type/ObjStm/Extends 50434 0 R>>stream
50436 0 obj<</First 2150/Length 1008/Filter/FlateDecode/N 200/Type/ObjStm/Extends 50434 0 R>>stream
50437 0 obj<</First 314/Length 236/Filter/FlateDecode/N 32/Type/ObjStm/Extends 50434 0 R>>stream
50438 0 obj<</First 17/Length 74/Filter/FlateDecode/N 2/Type/ObjStm>>stream
50439 0 obj<</First 2330/Length 6557/Filter/FlateDecode/N 200/Type/ObjStm>>stream
50440 0 obj<</First 2344/Length 6835/Filter/FlateDecode/N 200/Type/ObjStm/Extends 50439 0 R>>stream
50441 0 obj<</First 1726/Length 5368/Filter/FlateDecode/N 148/Type/ObjStm/Extends 50439 0 R>>stream
50442 0 obj<</First 1931/Length 4726/Filter/FlateDecode/N 200/Type/ObjStm>>stream
[…]
2165 0 obj<</Subtype/image#2Fjpeg/Length 41936/Filter/FlateDecode/DL 59244/Params<</CreationDat
e(D:20130604134130+05'30')/ModDate(D:20130604134130+05'30')/Size 59244/CheckSum(
13909 0 obj<</Subtype/Image/Length 2311108/Filter/DCTDecode/Name/X/BitsPerComponent 8/ColorSpac
e/DeviceRGB/Width 2211/DecodeParms<</Columns 2211/Blend 1/VSamples[1 1 1 1]/Rows 2787/HSamples[
1 1 1 1]/Colors 3/ColorTransform 1/QFactor 0.0>>/Height 2787/Type/XObject>>stream
14037 0 obj<</Length 287/Filter/FlateDecode>>stream
```

By comparison here's some from after the mutools conversion:
```
<</DecodeParms<</ColorTransform 1>>/Type/XObject/Subtype/Image/Filter/DCTDecode/BitsPerComponent 8/Width 2211/Height 2787/ColorSpace/DeviceRGB/Length 2311108>>
<</Length 49/Filter/FlateDecode>>
<</Length 324/Filter/FlateDecode>>
<</Length 287/Filter/FlateDecode>>
<</Length1 1545/Subtype/CIDFontType0C/Length 1391/Filter/FlateDecode>>
<</Length 455/Filter/FlateDecode>>
<</Length 324/Filter/FlateDecode>>
<</Length 354/Filter/FlateDecode>>
<</Length1 10493/Subtype/CIDFontType0C/Length 7803/Filter/FlateDecode>>
<</Length 287/Filter/FlateDecode>>
<</Length1 2137/Subtype/CIDFontType0C/Length 1992/Filter/FlateDecode>>
<</Length 1528/Filter/FlateDecode>>
<</Length 291/Filter/FlateDecode>>
<</Length1 8049/Subtype/CIDFontType0C/Length 6238/Filter/FlateDecode>>
```

**Comment by benoit-pierre** [comment]
It takes 3 seconds to load on my desktop (a 10 years old machine), and ~45 seconds on my Kindle Paperwhite (6th generation, ~8 years old I think?). That's with my version (MuPDF 1.23.0, ZLIB 1.23.0, among other updates).

**Comment by benoit-pierre** [comment]
Desktop is with MuPDF 1.23.2 and ZLIB 1.23.0. If zlib compression is involved, I know that using zlib-ng can speed up things (at least it does for rendering to PNGs).

**Comment by Frenzie** [comment]
> I know that using zlib-ng can speed up things 

The strange things to me are:

a. the output from `mutool convert -O compress etc` is also zlib (oops, I see I wrote glib above, luckily you understood me)
b. Poppler is fast

Certainly not an objection to zlib-ng (though is an old Kindle or Kobo sufficiently "NG"?), but MuPDF and Poppler share the same zlib on my system afaik, but I think the not quite investigation I shared above seems to point in another direction.

Anyway, I probably should've jumped straight to a profiler, heh.

**Comment by hugleo** [comment]
Approximate times on KDE Arch Linux:

okular ./Kimball_The-Data-Warehouse-Toolkit-3rd-Edition.pdf 0.6 s
mupdf(1.23.2) ./Kimball_The-Data-Warehouse-Toolkit-3rd-Edition.pdf 2.3 s
mupdf(1.21.0) ./Kimball_The-Data-Warehouse-Toolkit-3rd-Edition.pdf 2.3 s

strings mupdf | grep -i 'zlib'
ZLIB_1.2.0
zlib error: inflateInit2 failed
zlib error: inflateEnd: %s
ignoring zlib error: %s
zlib inflateInit2 error: %s
zlib inflate error: %s
zlib inflateEnd error: %s

**Comment by benoit-pierre** [comment]
@Frenzie : and it's similarly slow with KOReader on your desktop? You could give [this AppImage](https://github.com/benoit-pierre/koreader/suites/15802850658/artifacts/901813739) a try.

Still, 5min on a Kindle Paperwhite 3 vs 45s on a Kindle Paperwhite 2 means there are some improvements after updating some dependencies and/or optimizing the build.

**Comment by benoit-pierre** [comment]
@luutuankiet : do you have reflowing enabled by default?

**Comment by Frenzie** [comment]
> @Frenzie : and it's similarly slow with KOReader on your desktop? You could give [this AppImage](https://github.com/benoit-pierre/koreader/suites/15802850658/artifacts/901813739) a try.

I could, but I don't understand what you're asking. "It takes 3 seconds to load on my desktop (a 10 years old machine)" is my situation and my result.

**Comment by benoit-pierre** [comment]
I guess I interpreted "several seconds" as more than 3, at least 5! ;)

**Comment by Frenzie** [comment]
Well yes, the debug build is about 4 seconds, the release build of KOReader (and MuPDF) is about 3. Your AppImage is identical as expected.

**Comment by luutuankiet** [comment]
> @luutuankiet : do you have reflowing enabled by default?

@benoit-pierre no I didn't enable reflowing by default.

**Comment by hugleo** [comment]
On my Onyx device:
The first time I executed it took 12 seconds. 
Subsequent times took minutes or... I found that if I click on a KOReader blank side while it's still loading, the file loads immediately. So I think it's not a processing issue. It's more like an invisible "Please, press any key to continue..." :P

**Comment by hugleo** [comment]
It's because this PDF has an embedded image file, 9781118530771.jpg. I removed that from the PDF and it opened instantly.

**Comment by hugleo** [comment]
$ python script.py Kimball_The-Data-Warehouse-Toolkit-3rd-Edition.pdf > blocks_in_original
$ python script.py removed.pdf > blocks_in_removed

diff are the same
$ diff blocks_in_original blocks_in_removed 

It appears that a 3-second delay is applied during the first fitz.open operation and also some ms delay, each execution of load_page

[script.zip](https://github.com/koreader/koreader/files/12536822/script.zip)

**Comment by hugleo** [comment]
The 3s delay seems to be caused by a function with three nested loops

mupdf-1.23.3-source/source/pdf/pdf-xref.c
void pdf_xref_entry_map(fz_context *ctx, pdf_document *doc, void (*fn)(fz_context *, pdf_xref_entry *, int, pdf_document *, void *), void *arg) ...

		for (j = 0; j < doc->num_xref_sections; j++)
		{
			pdf_xref *xref = &doc->xref_sections[j];
			doc->xref_base = j;

			for (sub = xref->subsec; sub != NULL; sub = sub->next)
			{
				for (i = sub->start; i < sub->start + sub->len; i++)
				{
					pdf_xref_entry *entry = &sub->table[i - sub->start];
					if (entry->type)
						fn(ctx, entry, i, doc, arg);
					printf("hit\n");
				}
			}
		}


$  build/release/mupdf-x11  ~/Desktop/Kimball_The-Data-Warehouse-Toolkit-3rd-Edition.pdf > hit_in_original
$  build/release/mupdf-x11  ~/Desktop/removed.pdf > hit_in_removed
		
$ cat hit_in_original | wc -l
76352
$ cat hit_in_removed | wc -l
12464

For the original file we hit 76,352 times in the innermost loop
For the file with the embedded image removed we hit 12464 times in the innermost loop

**Comment by hugleo** [comment]
The issue definitely is not with the embedded file. It's just that the way I regenerated the PDF removed some other information that was causing the problem.

I used the following commands:

$ pdfdetach -saveall Kimball_The-Data-Warehouse-Toolkit-3rd-Edition.pdf
$ pdfattach my500pagefile.pdf 9781118530771.jpg embedded.pdf

and embedded.pdf opens instantaneously.

Using the command:

$ qpdf --decrypt Kimball_The-Data-Warehouse-Toolkit-3rd-Edition.pdf new.pdf

The embedded file remains inside and when opening new.pdf with the MuPDF desktop it opens instantaneous ;-)
